[libc++][ranges] P1223R5: `find_last`

strega-nil opened the pull request:
Implements [P1223R5][] completely.

Includes an implementation of `find_last`, `find_last_if`, and
`find_last_if_not`.

Note, this is my first contribution to libc++ in a very long time, so please keep that in mind when reviewing!

[P1223R5]: https://wg21.link/p1223r5

strega-nil committed:
P1223R5: find_last

Implements [P1223R5][] completely.

Includes an implementation of `find_last`, `find_last_if`, and
`find_last_if_not`.

[P1223R5]: https://wg21.link/p1223r5

Zingam commented:
We usually add a prefix to the title to indicate what a PR relates to, e.g. [libc++][ranges]. BTW I think there is already an open PRbut I think it is inactive
https://github.com/llvm/llvm-project/pull/67270

frederick-vs-ja commented on libcxx/include/__algorithm/ranges_find_last.h:
```diff
@@ -0,0 +1,103 @@
+
+  template <forward_iterator _Iter, sentinel_for<_Iter> _Sent, class _Type, class _Proj = identity>
+    requires indirect_binary_predicate<ranges::equal_to, projected<_Iter, _Proj>, const _Type*>
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI constexpr subrange<_Iter>
+  operator()(_Iter __first, _Sent __last, const _Type& __value, _Proj __proj = {}) const {
```
It seems that we can and should use `static operator()` for C++23 (and later)-only stateless functors. (Ditto below.)

Zingam commented:
You should also update the ranges status page: https://libcxx.llvm.org/Status/Ranges.html

strega-nil commented:
@Zingam 

> You should also update the ranges status page: https://libcxx.llvm.org/Status/Ranges.html

Isn't this done by updating `libcxx/docs/Status/Cxx23Papers.csv`?

Zingam commented:
> @Zingam
> 
> > You should also update the ranges status page: https://libcxx.llvm.org/Status/Ranges.html
> 
> Isn't this done by updating `libcxx/docs/Status/Cxx23Papers.csv`?

We have subprojects status pages also.

This is the file for the release notes (LLVM 19): `libcxx/docs/ReleaseNotes/19.rst`
This is the file for the ranges status page: `libcxx/docs/Status/Ranges.rst` and the table with the algorithms `libcxx/docs/Status/RangesAlgorithms.csv`

strega-nil committed:
code review comments

- add `static` to `operator()`
- update release notes

Zingam commented on libcxx/docs/FeatureTestMacroTable.rst:
```diff
@@ -360,6 +360,8 @@ Status
     ---------------------------------------------------------- -----------------
     ``__cpp_lib_ranges_contains``                              ``202207L``
     ---------------------------------------------------------- -----------------
+    ``__cpp_lib_ranges_find_last``                             ``202207L``
```
It seems you haven't run the generator script. Have you set this value manually? It gets set automatically.

Zingam replied:
Please check this list:
https://libcxx.llvm.org/Contributing.html#pre-commit-check-list

What about :)
  ninja -C "${build_dirpath}" libcxx-generate-files

strega-nil committed:
attempt to fix it?

strega-nil committed:
fixy fixy

cjdb commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.find.last/ranges.find_last.pass.cpp:
```diff
@@ -0,0 +1,294 @@
+        int comp;
+        int other;
+      };
+      S a[]    = {{0, 0}, {0, 2}, {0, 1}};
```
I'd prefer it if we used `std::array` outside of dedicated C array tests, please.

strega-nil replied:
I disagree; I find using `std::array` here unhelpful (it's just an extra template instantiation for no reason). If this was product code, I might agree, but it does not seem helpful to avoid C arrays in test code for the standard library. I'll certainly do it if it's necessary, but just wanted to register the disagreement.

cjdb commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.find.last/ranges.find_last.pass.cpp:
```diff
@@ -0,0 +1,294 @@
+  friend bool operator==(const Comparable& lhs, long long rhs) { return comparable_data[lhs.index_] == rhs; }
+};
+
+void test_deque() {
```
Can we parameterise this and test all the sequence containers, please?

strega-nil replied:
Yeah, I just copied these tests from `ranges::find`

strega-nil replied:
Actually, I'm not sure it's useful to test all the sequence containers - it seems mostly like wasted compute. How about just testing a forward, bidirectional, and random access container?

cjdb replied:
@EricWF once made an argument that testing that this works with all of our containers is good for ensuring all of our containers can use all of our algorithms without issue (I think the idea is that different algorithms will expose different usage patterns and potentially expose latent bugs). It's not integration-level, but it does achieve something more than generating heat.

strega-nil replied:
I mean, I will do it if others agree with you here, but it just seems very unhelpful to me - this algorithm doesn't do anything interesting, really. All it is is `operator--`, `operator++`, and `operator*`

cjdb commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.find.last/ranges.find_last_if.pass.cpp:
```diff
@@ -0,0 +1,263 @@
+//===----------------------------------------------------------------------===//
```
Would be good to have a `test_deque()`-equivalent in this file.

strega-nil replied:
I actually don't think it's necessary - I think it's just a leftover from when `ranges::find` was broken with `deque`. No reason to spend the compute cycles.

cjdb commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.find.last/ranges.find_last_if_not.pass.cpp:
```diff
@@ -0,0 +1,266 @@
+//===----------------------------------------------------------------------===//
```
Ditto.

cjdb reviewed (commented):
Great work! Thanks for this :)

Would you mind adding this to the release notes, please?

I'd also prefer it if we had one `find_last.h` instead of three files, but this is non-blocking, given precedent. An advantage to doing it in this commit is that the commit history would be more likely to be preserved should we consolidate files at a later point.

philnik777 commented on libcxx/include/__algorithm/ranges_find_last.h:
```diff
@@ -0,0 +1,103 @@
+namespace __find_last {
+struct __fn {
+  template <class _Iter, class _Sent, class _Type, class _Proj>
+  _LIBCPP_HIDE_FROM_ABI constexpr static subrange<_Iter>
```
Please don't just put `_LIBCPP_ALWAYS_INLINE` somewhere without good reason. There is no precedent for this in libc++.

strega-nil commented on libcxx/include/__algorithm/ranges_find_if.h:
```diff
@@ -14,6 +14,8 @@
 #include <__functional/invoke.h>
 #include <__functional/ranges_operations.h>
 #include <__iterator/concepts.h>
+#include <__iterator/next.h>
+#include <__iterator/prev.h>
```
Incorrect editing, apparently.

strega-nil commented on libcxx/include/__algorithm/ranges_find_last_if_not.h:
```diff
@@ -0,0 +1,70 @@
+            indirect_unary_predicate<projected<_Iter, _Proj>> _Pred>
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI constexpr static subrange<_Iter>
+  operator()(_Iter __first, _Sent __last, _Pred __pred, _Proj __proj = {}) {
+    auto __pred2 = [&__pred](auto&& __e) -> bool { return !std::invoke(__pred, std::forward<decltype(__e)>(__e)); };
```
`not_fn` does an extra copy of the function object. Could be worked around via `not_fn(ref(__pred))`. The reason I did this wsa that this is what `ranges::find_last_if_not` does.

cjdb replied:
Oh right, it copies. Nothing to change then.

cjdb commented:
> We usually add a prefix to the title to indicate what a PR relates to, e.g. [libc++][ranges]. BTW I think there is already an open PR but I think it is inactive. We usually ask on Discord to avoid duplicated work #67270
> 
> Please add a release note also. https://libcxx.llvm.org/ReleaseNotes.html

That's my fault: I'd forgotten about that PR when suggesting the work to @strega-nil. Having said that, #67270 hasn't been touched in a long while.

ldionne commented:
@strega-nil I closed #67270 as inactive. Please feel free to take tests and other material from that PR (if you do, please credit the original author of the patch in your commit message). This is now the canonical PR for `find_last` & friends.

strega-nil committed:
(wip) merge ranges_find_last*.h into ranges_find_last.h

Additionally, make __find_last_impl into one function for all three

Finally, change the tests for find_last to see if people are good with that version
(hopefully, yes, in which case I'll do the rest)

strega-nil commented:
Pinging reviewers for a re-review - I have completed the merge of the files, and I've modified how the `find_last` test works - if this modification of the test works for y'all, I'll make the same changes to the other tests and say it's shippable :)

strega-nil committed:
fix tests

strega-nil committed:
finish off the tests

strega-nil committed:
format it

ldionne commented on libcxx/docs/ReleaseNotes/19.rst:
```diff
@@ -55,6 +55,7 @@ Implemented Papers
 - P2231R1 - Missing ``constexpr`` in ``std::optional`` and ``std::variant``
 - P0019R8 - ``std::atomic_ref``
 - P2389R2 - Alias template ``dims`` for the ``extents`` of ``mdspan``
+- P1223R5 - ``find_last``
```
```suggestion
- P1223R5 - ``ranges::find_last()``, ``ranges::find_last_if()``, and ``ranges::find_last_if_not()``
```

ldionne commented on libcxx/include/__algorithm/ranges_find_last.h:
```diff
@@ -0,0 +1,175 @@
+namespace ranges {
+
+template <class _Iter, class _Sent, class _Pred, class _Proj>
+_LIBCPP_HIDE_FROM_ABI constexpr static subrange<_Iter>
```
```suggestion
_LIBCPP_HIDE_FROM_ABI constexpr subrange<_Iter>
```

This shouldn't be static, it's a template at file scope!

ldionne commented on libcxx/include/__algorithm/ranges_find_last.h:
```diff
@@ -0,0 +1,175 @@
+  struct __op {
+    const _Type& __value;
+    template <class _Elem>
+    constexpr decltype(auto) operator()(_Elem&& __elem) const {
```
```suggestion
    _LIBCPP_HIDE_FROM_ABI constexpr decltype(auto) operator()(_Elem&& __elem) const {
```

ldionne commented on libcxx/include/__algorithm/ranges_find_last.h:
```diff
@@ -0,0 +1,175 @@
+  struct __op {
+    _Pred& __pred;
+    template <class _Elem>
+    constexpr decltype(auto) operator()(_Elem&& __elem) const {
```
```suggestion
    _LIBCPP_HIDE_FROM_ABI constexpr decltype(auto) operator()(_Elem&& __elem) const {
```

ldionne commented on libcxx/include/__algorithm/ranges_find_last.h:
```diff
@@ -0,0 +1,175 @@
+  struct __op {
+    _Pred& __pred;
+    template <class _Elem>
+    constexpr decltype(auto) operator()(_Elem&& __elem) const {
```
```suggestion
    _LIBCPP_HIDE_FROM_ABI constexpr decltype(auto) operator()(_Elem&& __elem) const {
```

ldionne commented on libcxx/include/__algorithm/ranges_find_last.h:
```diff
@@ -0,0 +1,175 @@
+namespace __find_last_if {
+struct __fn {
+  template <class _Pred>
+  struct __op {
```
Actually, is there any reason for not using a lambda (here and elsewhere)?

strega-nil replied:
I feel it's easier to read, plus only once instead of in both functions.

ldionne commented on libcxx/include/__algorithm/ranges_find_last.h:
```diff
@@ -0,0 +1,175 @@
+    return subrange<_Iter>(__first, __first);
+  }
+
+  if constexpr (bidirectional_iterator<_Iter>) {
```
In case of an iterator that is bidirectional but where `ranges::next(iterator, sentinel)` is `O(N)`, this is potentially wasteful. We will first walk all the way to `__last` inside `ranges::next`, and then walk back until we find something that satisfies the predicate.

Writing this, I realize that the alternative is not necessarily better -- if you search from the start, then you potentially call the predicate all the way from the beginning. This depends on whether we expect the predicate to generally match near the end of the sequence or not.

I think it's a fair assumption to optimize this for finding stuff near the end of the sequence. So no change required, I was just thinking out loud.

strega-nil replied:
Yeah, this was my exact thought process.

ldionne commented on libcxx/include/algorithm:
```diff
@@ -102,6 +102,26 @@ namespace ranges {
     constexpr borrowed_iterator_t<R>
       find_if_not(R&& r, Pred pred, Proj proj = {});                                                      // since C++20
 
+  template<forward_iterator I, sentinel_for<I> S, class T, class Proj = identity>
+    requires indirect_binary_predicate<ranges::equal_to, projected<I, Proj>, const T*>
+    constexpr subrange<I> find_last(I first, S last, const T& value, Proj proj = {});                     // since C++23
+  template<forward_range R, class T, class Proj = identity>
+    requires
+      indirect_binary_predicate<ranges::equal_to, projected<iterator_t<R>, Proj>, const T*>
+    constexpr borrowed_subrange_t<R> find_last(R&& r, const T& value, Proj proj = {});                   // since C++23
+  template<forward_iterator I, sentinel_for<I> S, class Proj = identity,
+           indirect_unary_predicate<projected<I, Proj>> Pred>
+    constexpr subrange<I> find_last_if(I first, S last, Pred pred, Proj proj = {});                      // since C++23
+  template<forward_range R, class Proj = identity,
+           indirect_unary_predicate<projected<iterator_t<R>, Proj>> Pred>
+    constexpr borrowed_subrange_t<R> find_last_if(R&& r, Pred pred, Proj proj = {});                     // since C++23
+  template<forward_iterator I, sentinel_for<I> S, class Proj = identity,
+           indirect_unary_predicate<projected<I, Proj>> Pred>
+    constexpr subrange<I> find_last_if_not(I first, S last, Pred pred, Proj proj = {});                  // since C++23
+  template<forward_range R, class Proj = identity,
+           indirect_unary_predicate<projected<iterator_t<R>, Proj>> Pred>
+    constexpr borrowed_subrange_t<R> find_last_if_not(R&& r, Pred pred, Proj proj = {});                 // since C++23
```
A few newlines would make this easier to read.

strega-nil replied:
It is just as hard to read in the standard, but definitely can do 😝

ldionne commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.find.last/ranges.find_last.pass.cpp:
```diff
@@ -0,0 +1,206 @@
+};
+
+constexpr bool test() {
+  types::for_each(types::type_list<char, wchar_t, int, long, TriviallyComparable<char>, TriviallyComparable<wchar_t>>{},
```
Any specific reason for using `wchar_t`? You'd have to make that conditional on `TEST_HAS_NO_WCHAR_T`.

strega-nil replied:
Copied from the test for `find`

ldionne commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.find.last/ranges.find_last.pass.cpp:
```diff
@@ -0,0 +1,206 @@
+  ElementT el_;
+
+public:
+  TEST_CONSTEXPR TriviallyComparable(ElementT el) : el_(el) {}
```
```suggestion
  constexpr TriviallyComparable(ElementT el) : el_(el) {}
```

ldionne commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.find.last/ranges.find_last.pass.cpp:
```diff
@@ -0,0 +1,206 @@
+#include <algorithm>
+#include <array>
+#include <cassert>
+#include <deque>
```
You include `deque` here but don't use it. Did you intend to add a test with deque?

strega-nil replied:
Old test from the copy from `find`.

ldionne commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.find.last/ranges.find_last_if.pass.cpp:
```diff
@@ -0,0 +1,258 @@
+struct add_const_to_ptr<T*> {
+  using type = const T*;
+};
+template <class T>
+using add_const_to_ptr_t = add_const_to_ptr<T>::type;
```
```suggestion
template <class T>
using add_const_to_ptr_t = std::add_pointer_t<std::add_const_t<std::remove_pointer_t<T>>>;
```

Non-blocking, just a suggestion to simplify things a bit. Would also apply to the other test files.

strega-nil replied:
Great suggestion.

ldionne reviewed (changes requested):
Thanks, this is a nice patch! I have a few comments but I don't see any major issues.

strega-nil committed:
ldionne's recommendations

strega-nil committed:
format dangit

strega-nil commented:
@ldionne I think I've fixed all of your comments :)

ldionne reviewed (approved):
LGTM, thanks for the quick turnaround!
